uni/ulsp/completion.icn

```diff
@@ -0,0 +1,207 @@
+package ulsp
+
+import json
+
+class CompletionHandler(
+	workspace,
+	params
+)
+
+	method setVariables(new_workspace, new_params)
+		workspace := new_workspace
+		params := new_params
+	end
+
+	method run()
+		local _context, line, character, results_table, results
+		
+		line := params["position"]["line"]
+		character := params["position"]["character"]
+
+		_context := getContext(line, character, workspace.temp_file)
+
+		results_table := table()
+		results_table["isComplete"] :=  "__true__"
+		results_table["items"] := []
+
+		case _context of {
+			"inside comment" : {
+				results_table["items"] := []
+			}
+			"after dot operator" : {
+				desired_line := getDesiredLine(line, workspace.temp_file)	
+				objectName := getObjectName(desired_line, character) 
+				workspace.buildObjectCompletionItems(results_table, objectName)
+			}
+			default : { 
+				buildDefaultCompletionItems(results_table, workspace)
+			}
+		}
+
+		results := tojson(results_table)
+		return results
+
+
+	end
+
+
+	method getObjectName(line, character)
+		identifiers := &letters ++ &digits ++ '_' ++ '-'
+
+		line ? {
+			temp := tab(character + 1)
+			while c := move(-1) do {
+				if (c ** identifiers) ~== c then { 
+					break
+				}
+			}
+
+			if ="." then {
+				move(-1)
+
+				while c := move(-1) do {
+					if (c ** identifiers) ~== c then { 
+						move(1)
+						break
+					}
+				}
+				objectName := tab(upto("."))
+				return objectName
+				
+			}
+		}
+
+	end
+
+	method getDesiredLine(line, file)
+		f := open(file)
+		every !line do read(f)
+		desired_line := read(f)
+		return desired_line
+	end
+
+	method getContext(line, character, file)
+
+		desired_line := getDesiredLine(line, file)
+
+		#-----------------------------------------------#
+		#----------- Case "start of file" --------------#
+		#-----------------------------------------------#
+
+		if desired_line = &null return &null
+
+		#-----------------------------------------------#
+		#----------- Case "inside comment" -------------#
+		#-----------------------------------------------#
+
+		desired_line ? {
+			while (&pos < character) do {
```

Comment:
Why loop? can't we just use `upto()`?

desired_line ? if upto('#') then return "inside comment"